text/0000-dyno.md

```diff
@@ -0,0 +1,294 @@
+
+`provide_any` could live in its own crate, rather than in libcore. However, this would not be useful for `Error`.
+
+`provide_any` could be a module inside `any` rather than a sibling (it could then be renamed to `provide` or `provider`).
+
+There are numerous ways to tweak the API of a module like `provide_any`. The dyno and object-provider crates provide two such examples. There are many others, for example providing more patterns of types without requiring tags, not providing any common type patterns (i.e., always requiring tags), not exposing tags at all, etc.
+
+One hazard that must be avoided with such tweaks is that the API must distinguish between reference types with a `'static` lifetime and value types (with no lifetimes), either by using type tags or by having separate mechanisms for handling references and values. If this is done improperly, the API could be unsound. As an example, consider an API which lets an object provide a reference with type `&'a str` (where `'a` is the lifetime of some local scope), then a caller requests an object of type `&'static str` using an API designed for values (possible because that type satisfies the `'static` bound). Without some care, it is possible for the two types to be confused (because the lifetime is not reflected in the type tag or `TypeId`) and for the `&'a str` to be returned with the wrong lifetime. I believe this is not possible in the current proposal, but was possible in an earlier, more ergonomic iteration.
+
+A slightly different approach is to remove the convenience methods and always require the use of type tags. I believe this is sub-optimal since it requires users to learn a new concept (type tags) and makes using the methods less ergonomic. However, it does have advantages: code is less prone to silent inference errors since the type tag must always be explicitly specified, and the API is smaller.
+
+# Prior art
+[prior-art]: #prior-art
+
+
+Operations involving runtime types are intrinsically tied to the specifics of a language, its runtime, and type system. Therefore, there is not much prior art from other languages.
+
+A closely related feature from other languages is reflection. However, reflective field access is usually name-driven rather than type-driven. Due to Rust's architecture, general reflection is impossible.
+
+In Rust, there are several approaches to the problem. This proposal is adapted from [dyno](https://github.com/mystor/dyno/); [object provider](https://github.com/mystor/object-provider/) is a similar crate from the same author.
+
+Some Rust error libraries provide roughly similar functionality. For example, [Anyhow](https://github.com/dtolnay/anyhow) (and [Eyre](https://github.com/yaahc/eyre)) allow adding context to errors in `Result`s, which can be accessed by downcasting the error object to the context object.
+
+
+# Unresolved questions
```

Comment:
Bridging into `async` - how can I ask for an `impl Future<Output = std::error::Backtrace>`?

Reply:
I'm not sure if you can get a monomorphized generic future type out of this where the type is callee provided but you should be able to pretty easily request a `&dyn Future<Output = Backtrace>` or `Box<dyn Future<Output = Backtrace>>`, and if you know the concrete type of the `impl Future` you're interested in you can always request that directly.